ramips: add support for TP-Link Archer A6 v3

vrpatil opened the pull request:
Patch to add support for the TP-Link Archer A6 v3.

The device is sold in US and India (FCC ID TE7A6V3) with following specifications.


### **Specification**
- MediaTek MT7621 SOC -  880MHz 
- RAM:               128MB DDR3
- SPI Flash:        W25Q128 (16MB)
- Ethernet:          MT7530  5x 1000Base-T
- WiFi 5GHz:      Mediatek MT7613BE 
- WiFi 2.4GHz:   Mediatek MT7603E
- UART/ Serial:  115200 8n1

### How to Use
**Flashing from TP-Link Web Interface**
- Go to "Advanced/System Tools/Firmware Update".
- Click "Browse" and upload the OpenWrt factory image: openwrt-ramips-mt7621-tplink_archer-a6-v3-squashfs-factory.bin.
- Click the "Upgrade" button, and select "Yes" when prompted.


**TFTP Booting**
- Setup TFTP server with address 192.168.0.5
- While starting U-boot press '4' key to stop autoboot. 
- Copy the openwrt-ramips-mt7621-tplink_archera6-v3-initramfs-kernel.bin to TFTP server folder and rename it as test.bin
- From u-boot command prompt run _tftpboot_ followed by _bootm_

### Recovery 
Archer A6 V3 has recovery page activated if SPI booting from flash fails. Recovery page can also be activated from serial console, by pressing 'x' while u-boot is starting. But TFTP boot can be activated only from u-boot serial console.
Recovery address: 192.168.0.1

The OEM Firmware can be downloaded from: https://www.tp-link.com/us/support/download/archer-a6/

Signed-off-by: Vinay P post2vinay@gmail.com

adschm commented:
This has severe formal issues. Please visit https://openwrt.org/submitting-patches again and check how other device support patches in this target look like from git history.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1102,6 +1102,20 @@ define Device/tplink_re650-v1
 endef
 TARGET_DEVICES += tplink_re650-v1
 
+define Device/tplink_archera6-v3
```
alphabetic sorting ...

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1102,6 +1102,20 @@ define Device/tplink_re650-v1
+define Device/tplink_archera6-v3
+  $(Device/dsa-migration)
+  $(Device/tplink-safeloader)
+  DEVICE_MODEL := ArcherA6
```
missing space after Archer

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1102,6 +1102,20 @@ define Device/tplink_re650-v1
+  DEVICE_VARIANT := v3
+  DEVICE_PACKAGES := kmod-mt7603 kmod-mt7615e \
+	kmod-mt7663-firmware-ap kmod-mt7663-firmware-sta
+  TPLINK_BOARD_ID := ArcherA6-V3
```
IIRC it's common to have this variable only-upper-case

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1102,6 +1102,20 @@ define Device/tplink_re650-v1
+  TPLINK_BOARD_ID := ArcherA6-V3
+  KERNEL := $(KERNEL_DTB) | uImage lzma  
+  IMAGE_SIZE := 6016k
+  SUPPORTED_DEVICES += archera6-v3
```
This line can be dropped.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -85,7 +85,8 @@ tplink,re350-v1)
 tplink,re500-v1|\
-tplink,re650-v1)
+tplink,re650-v1|\
+tplink,archera6-v3)
```
Sorting ...

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -33,6 +33,9 @@ ramips_setup_interfaces()
 		ucidef_set_interface_lan "lan"
 		;;
+	tplink,archera6-v3)
+		ucidef_set_interfaces_lan_wan "eth1 eth2 eth3 eth4" "wan"
```
You should use lanX here and then may just use the default case ...

adschm commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -1058,7 +1058,40 @@ static struct device_info boards[] = {
+			"SupportList:\n"
+			"{product_name:Archer A6,product_ver:3.0.0,special_id:55530000}\n"
+			"{product_name:Archer A6,product_ver:3.0.0,special_id:54570000}\n"
+			"{product_name:Archer C6,product_ver:3.0.0,special_id:55530000}\n",
```
That's not A6, that's C6. If you want to support both devices, this should be mentioned somewhere, or one should even have two images for them.

adschm commented on target/linux/ramips/dts/mt7621_tplink_archera6-v3.dts:
```diff
@@ -0,0 +1,204 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "tplink,archera6-v3", "mediatek,mt7621-soc";
```
compatible should be "tplink,archer-a6-v3" ...

adschm commented on target/linux/ramips/dts/mt7621_tplink_archera6-v3.dts:
```diff
@@ -0,0 +1,204 @@
+
+			/* range 0xe50000 to 0xff0000 is empty in vendor
+			 * firmware, so we do not use it either
+			 */
```
Your partitioning can't be right. Having only a 0x200000 firmware partition also tells me that this has never been run on device.

adschm commented on target/linux/ramips/dts/mt7621_tplink_archera6-v3.dts:
```diff
@@ -0,0 +1,204 @@
+		port@0 {
+			status = "okay";
+			label = "wan";
+			mtd-mac-address = <&config 0x8>;
```
Where does the 0x8 come from suddenly here?

vrpatil replied:
@musashino205 confirm the MAC address is stored at 0x08 offset of `config` partition. But it is not read and applied by OpenWrt.

musashino205 replied:
The MAC address in eeprom data (offset: ```0x4```) in "Factory" partition cannot be applied if the device has a MT7663 (MT7613), but I don't know about "config" partition...

vrpatil commented:
@adschm  Will submit a new patch/merge request. Pull requests from upstream/master are appearing as commits in my branch.

adschm commented:
No, please do not submit a new request, but update this one properly (git push -f ...)

musashino205 commented:
This is a personal question, is a correct MAC address assigned to MT7613BEN on OpenWrt?
I'm working to add support for a router that have a MT7613BEN, but a random MAC address is assigned to it. The correct MAC address in eeprom data on the flash is ignored.

vrpatil commented:
@musashino205 yes I also have same issue. It is taking random MAC address. Not sure the address pointed in the config partition is correct one. Haven't tried to finding the right address in config partition.

musashino205 commented:
@vrpatil OK, I tried to add debugging outputs, the eeprom data in flash is not used by mt7615e driver for mt7663 (mt7613).

changes:

```
diff --git a/mt7615/eeprom.c b/mt7615/eeprom.c
index adc797d..e943d36 100644
--- a/mt7615/eeprom.c
+++ b/mt7615/eeprom.c
@@ -334,9 +334,11 @@ int mt7615_eeprom_init(struct mt7615_dev *dev, u32 addr)
 
 	ret = mt7615_check_eeprom(&dev->mt76);
 	if (ret && dev->mt76.otp.data) {
+		dev_info(dev->mt76.dev, "use only otp data for mt7615e driver\n");
 		memcpy(dev->mt76.eeprom.data, dev->mt76.otp.data,
 		       MT7615_EEPROM_SIZE);
 	} else {
+		dev_info(dev->mt76.dev, "use eeprom data for mt7615e driver\n");
 		dev->flash_eeprom = true;
 		mt7615_cal_free_data(dev);
 	}
```

log:
```
[   14.307713] mt7615e 0000:02:00.0: enabling device (0000 -> 0002)
[   14.328191] mt7615e 0000:02:00.0: use only otp data for mt7615e driver
[   14.341312] mt7615e 0000:02:00.0: Invalid MAC address, using random address 5e:28:16:14:3c:b2
```

musashino205 commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -33,6 +33,9 @@ ramips_setup_interfaces()
 		;;
+	tplink,archer-a6-v3)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"
+		;;
```
This is the same configuration as default, so can be dropped.

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+		};
+
+		led_wan_orange {
+			label = "orange:led_wan";
```
orange:wan ?

vrpatil replied:
yes it has orange LED for wan. Not sure how to make use of it.

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		led_wan_orange {
```
wan_orange

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		led_lan {
```
lan

vrpatil replied:
used `led_` prefix to avoid namespace conflict with `lan`, `wan` used in ETH definitions, If without `led_*` is convention followed then can change to that.

adschm replied:
These are node names, not labels, they only exist within the node "leds", so there is no collision. That's why I typically request led_ prefix for the labels, ending up with the following where a label is required, e.g.:
`led_system: system {`

vrpatil replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+		};
+
+		led_lan {
+			label = "green:led_lan";
```
green:lan

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+			linux,default-trigger = "phy0tpt";
+		};
+
+		led_wan {
```
wan_green

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+		};
+
+		led_wan {
+			label = "green:led_wan";
```
green:wan

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+				label = "config";
+				reg = <0xfa0000 0x50000>;
+				read-only;
+			};
```
Add empty lines between nodes.

adschm replied:
Why is this resolved? I don't see empty lines here?

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,186 @@
+			status = "okay";
+			label = "lan4";
+		};
+		
```
Add empty lines between nodes and remove this empty line at the end.

adschm reviewed (commented):
This has formal issues as well, commit title needs to be fixed, commit message and Signed-off-by added.

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,181 @@
+		};
+
+		wan_orange {
+			label = "orange:wan_orange";
```
orange:wan

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,181 @@
+			linux,default-trigger = "phy0tpt";
+		};
+
+		wan {
```
wan_green

adschm commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,181 @@
+		port@3 {
+			status = "okay";
+			label = "lan3";
+		};
```
Please add empty lines between nodes.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -78,6 +78,12 @@ netgear,r6800)
+	ucidef_set_led_netdev "wifi2g" "Wifi 2.4G" "green:wifi2G" "wlan0"
+	ucidef_set_led_netdev "wifi5g" "Wifi 5G" "green:wifi5G" "wlan1"
+	ucidef_set_led_netdev "led_lan" "LAN act" "green:lan" "br-lan" "tx rx"
+	ucidef_set_led_netdev "led_wan" "WAN link" "green:wan" "wan" "link tx rx"
```
Well, this is not link, but link+activity.
Is there a specific reason why you use activity only for one and want to use link for the other?

vrpatil replied:
Tried to use link for br-lan didnt work. So kept it only for WAN

adschm replied:
Please don't mark everything as resolved just after you answered, I really hate it if I have to open everything again when looking at a PR.
What do you mean by "didnt work". Is it always off or always on?

vrpatil replied:
sorry about that. Never switches ON to indicate link status. For example even a LAN cable is connected to lanX, br-lan shows only activity status.

adschm replied:
Okay, that's what I would have expected. But then, we can just take the default and use
```
	ucidef_set_led_netdev "lan" "LAN" "green:lan" "br-lan"
	ucidef_set_led_netdev "wan" "WAN" "green:wan" "wan"
```

vrpatil replied:
Little confused. Does it mean  the default statements in switch case statement sufficient? There are not default ones present 01_led.sh, also wan LED node is now named as wan_green.. 
OR 
remove last part of the command like "tx rx" and keep only 4 arguments?

adschm replied:
```
tplink,archer-a6-v3)
	ucidef_set_led_netdev "lan" "LAN" "green:lan" "br-lan"
	ucidef_set_led_netdev "wan" "WAN" "green:wan" "wan"
	;;
```

That's what I want, and that's what all the others do.

The default for the fifth argument is "link rx tx".

vrpatil replied:
done!. Thanks for the feedback.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -78,6 +78,12 @@ netgear,r6800)
 	;;
+tplink,archer-a6-v3)
+	ucidef_set_led_netdev "wifi2g" "Wifi 2.4G" "green:wifi2G" "wlan0"
+	ucidef_set_led_netdev "wifi5g" "Wifi 5G" "green:wifi5G" "wlan1"
```
wifi LEDs can be dropped here, this is already done via triggers in DTS.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/01_leds:
```diff
@@ -78,6 +78,10 @@ netgear,r6800)
 	;;
+tplink,archer-a6-v3)
+	ucidef_set_led_netdev "lan" "LAN act" "green:lan" "br-lan" 
+	ucidef_set_led_netdev "wan" "WAN link act" "green:wan" "wan" 
```
Can you **please** finally remove the "act" here ...

vrpatil replied:
done!.

svanheule commented:
The ramips reset patch has been [merged](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=69551a244292f5baf0aee205ec6b891a03f1ee15), so that should solve ethernet not coming online when booting from flash.

The MT7613BE radio still won't work properly. That will probably need to come later via the mt76 driver, but at least the device won't be bricked after a factory install any more.

vrpatil commented:
@svanheule  thanks for the patch. yes, device shouldn't brick any more, and  5GHz radio isn't a blocker.

vrpatil commented:
@adschm are the changes fine? Let me know if any comments are pending closure.

svanheule commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,187 @@
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
```
You should add another empty line after aliases{ };

svanheule commented on target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:
```diff
@@ -0,0 +1,187 @@
+
+&ethernet {
+        pinctrl-names = "default";
+        pinctrl-0 = <&rgmii1_pins &mdio_pins>;
```
Are these pinctrl properties a remnant from trying to get the switch to work when booting from flash? I didn't add them on the EAP235-Wall, and they don't appear to be added on any MT7621 device.

vrpatil replied:
yes, will try removing and check if they are really making any sense.

vrpatil commented:
@adschm Can you please review the changes and let me know?

hbriese commented:
What's preventing this from being merged?

GChicha commented:
I tested this in archer C6 v3.2 and it looks to work just fine, however, I was unable to install using the tp-link web interface, I needed to boot using TFTP. I had to disable failsafe too, it seems to identify that I was pressing the WPS key.

AravindhStanley commented:
@adschm is there any specific reason this pull request is not merged. I could really use the support for this device.

frankis5 commented:
> 
> 
> @musashino205 yes same here It is taking random MAC address. Not sure the address pointed in the config partition is correct one. Haven't tried to finding the right address in config partition.

The offset should be at 0x8 and not at 0x10008 in the config partition to get it to read properly. Also if the assigned MAC addresses are to match the OEM ones, I believe LAN should be the MAC on the label, WAN should be +1, 2.4g wifi should be -1 (i.e., mtd-mac-address-increment = <(-1)> and 5g wifi should be -2.

vrpatil commented:
@ynezz or @adschm  can anyone please look into the PR.

frankis5 commented:
> 
> 
> @ynezz or @adschm can anyone please look into the PR.

The random MAC address issue still needs fixing It still says:

mtd-mac-address = <&config 0x10008>;

It should be:
mtd-mac-address = <&config 0x8>;

I've tested the 0x8 offset with the latest version of the PR and it fixes the random MAC address issue on the A6 V3.

vrpatil committed:
ramips: add support for TP-Link Archer A6 v3

The patch adds support for the TP-Link Archer A6 v3
The router is sold in US and India with FCC ID TE7A6V3

Specification
-------------
MediaTek MT7621 SOC
RAM:         128MB DDR3
SPI Flash:   W25Q128 (16MB)
Ethernet:    MT7530 5x 1000Base-T
WiFi 5GHz:   Mediatek MT7613BE
WiFi 2.4GHz: Mediatek MT7603E
UART/Serial: 115200 8n1

Device Configuration & Serial Port Pins
---------------------------------------
ETH Ports:    LAN4 LAN3 LAN2 LAN1 WAN
             _______________________
             |                     |
Serial Pins: |   VCC GND TXD RXD   |
             |_____________________|

LEDs:         Power Wifi2G Wifi5G LAN WAN

Build Output
------------
The build will generate following set of files
[1] openwrt-ramips-mt7621-tplink_archer-a6-v3-initramfs-kernel.bin
[2] openwrt-ramips-mt7621-tplink_archer-a6-v3-squashfs-factory.bin
[3] openwrt-ramips-mt7621-tplink_archer-a6-v3-squashfs-sysupgrade.bin

How to Use - Flashing from TP-Link Web Interface
------------------------------------------------
* Go to "Advanced/System Tools/Firmware Update".
* Click "Browse" and upload the OpenWrt factory image: factory.bin[2]
* Click the "Upgrade" button, and select "Yes" when prompted.

TFTP Booting
------------
Setup a TFTP boot server with address 192.168.0.5.
While starting U-boot press '4' key to stop autoboot.
Copy the initramfs-kernel.bin[1] to TFTP server folder, rename as test.bin
From u-boot command prompt run tftpboot followed by bootm.

Recovery
--------
Archer A6 V3 has recovery page activated if SPI booting from flash fails.
Recovery page can be activated from serial console only.
Press 'x' while u-boot is starting
Note: TFTP boot can be activated only from u-boot serial console.
Device recovery address: 192.168.0.1

Thanks to: Frankis for Randmon MAC address fix.

Signed-off-by: Vinay Patil <post2vinay@gmail.com>

vrpatil commented:
> > @ynezz or @adschm can anyone please look into the PR.
> 
> The random MAC address issue still needs fixing It still says:
> 
> mtd-mac-address = <&config 0x10008>;
> 
> It should be:
> mtd-mac-address = <&config 0x8>;
> 
> I've tested the 0x8 offset with the latest version of the PR and it fixes the random MAC address issue on the A6 V3.

Thanks for pointing out. Included the fix in the PR. Let me know if any issue.

sandmail32 commented:
It is unfortunate that this merge is taking time.
I am having some wierd issue with pppoe connection with the @vrpatil's branch.
so i created a [patch](https://gist.github.com/sandmail32/67d9f5c64a8979a4035278ef320a08a8) file from his repo...
apply this patch on master.
`$ git am < ~/Downloads/0001-ramips-add-support-for-TP-Link-Archer-A6-v3.patch`
and follow normal build instructions.

better than [these](https://stackoverflow.com/questions/1425892/how-do-you-merge-two-git-repositories) [weird](https://stackoverflow.com/questions/21353656/merge-git-repo-into-branch-of-another-repo/21353836) [commands](https://stackoverflow.com/questions/38468700/how-to-merge-branches-in-2-different-repositories). feeling lazy to follow that.

kpoman commented:
> Hey I have an archer c6 v3.2 . If this gets merged would it work for this too because the specs are almost identical?

Same here ! Would love to test and feed back if necessary.

kpoman commented:
Could anyone else review to allow this to be merged ?

blocktrron commented:
Merged to my staging tree with changes.

There were various whitespace issues with your DTS file. Please check this the next time using checkpatch:

```
ERROR: trailing whitespace
#86: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:11:
+^I$

ERROR: trailing whitespace
#94: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:19:
+^I$

ERROR: trailing whitespace
#97: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:22:
+^I};^I$

ERROR: trailing whitespace
#172: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:97:
+^I^I^I$

WARNING: DT compatible string "denx,uimage" appears un-documented -- check /home/dbauer/git/linux/Documentation/devicetree/bindings/
#175: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:100:
+				compatible = "denx,uimage";

ERROR: trailing whitespace
#178: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:103:
+^I^I^I$

ERROR: trailing whitespace
#184: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:109:
+^I^I^I$

ERROR: code indent should use tabs where possible
#202: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:127:
+        pinctrl-names = "default";$

WARNING: please, no spaces at the start of a line
#202: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:127:
+        pinctrl-names = "default";$

ERROR: code indent should use tabs where possible
#203: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:128:
+        pinctrl-0 = <&rgmii1_pins &mdio_pins>;$

WARNING: please, no spaces at the start of a line
#203: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:128:
+        pinctrl-0 = <&rgmii1_pins &mdio_pins>;$

ERROR: trailing whitespace
#242: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:167:
+^I^I$

ERROR: trailing whitespace
#247: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:172:
+^I^I$

ERROR: trailing whitespace
#252: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:177:
+^I^I$

ERROR: trailing whitespace
#257: FILE: target/linux/ramips/dts/mt7621_tplink_archer-a6-v3.dts:182:
+^I^I$

```

naks110 commented:
@vrpatil Can you guys please tell me what RAM chip Archer C6/A6 v3.20 uses? I will add the missing info on WikiDevi & others.
